src/journal/JournalMetadata.cc

```diff
@@ -1022,6 +1030,79 @@ void JournalMetadata::handle_notified(int r) {
   ldout(m_cct, 10) << "notified journal header update: r=" << r << dendl;
 }
 
+void JournalMetadata::disconnect_laggy_clients() {
+  assert(m_lock.is_locked());
+
+  ldout(m_cct, 20) << __func__ << dendl;
+
+  if (m_client_object_sets_behind_max <= 0) {
+    return;
+  }
+
+  std::queue<std::string> clients;
+  for (auto &c : m_registered_clients) {
+    if (c.id == m_client_id ||
+	c.state == cls::journal::CLIENT_STATE_DISCONNECTED) {
+      continue;
+    }
+    uint64_t object_set = 0;
+    if (!c.commit_position.object_positions.empty()) {
+      auto position = *(c.commit_position.object_positions.begin());
+      object_set = position.object_number / m_splay_width;
+    }
+    if (m_active_set > object_set + m_client_object_sets_behind_max) {
+      clients.push(c.id);
+    }
+  }
+
+  if (clients.empty()) {
+    ldout(m_cct, 20) << __func__ << ": no laggy clients to disconnect" << dendl;
+    return;
+  }
+
+  Context *ctx = new C_NotifyUpdate(this, nullptr);
+
+  while (clients.size() > 1) {
+    auto &client_id = clients.front();
+
+    ctx = new FunctionContext([this, client_id, ctx](int r1) {
+	ldout(m_cct, 1) << __func__ << ": " << client_id
+		        << ": flagging disconnected" << dendl;
+
+	librados::ObjectWriteOperation op;
+	client::client_update_state(&op, client_id,
+				    cls::journal::CLIENT_STATE_DISCONNECTED);
+
+	librados::AioCompletion *comp =
+	librados::Rados::aio_create_completion(ctx, nullptr,
+					       utils::rados_ctx_callback);
+	int r = m_ioctx.aio_operate(m_oid, comp, &op);
+	assert(r == 0);
+	comp->release();
+      });
+
+    clients.pop();
+  }
+
+  assert(clients.size() == 1);
```

Comment:
Oh -- I missed the FunctionContext.  Maybe move the bulk librados code section to a helper method so you can re-use it.  You also should be able to just keep adding ```client::client_update_state``` ops to a single ```librados::ObjectWriteOperation``` since each will update different omap keys.